pymeasure/adapters/prologix.py

```diff
@@ -84,37 +99,43 @@ def set_defaults(self):
     def ask(self, command):
         """ Ask the Prologix controller, include a forced delay for some instruments.
 
+        .. deprecated:: 0.10
+           Call `Instrument.ask` instead.
+
         :param command: SCPI command string to be sent to instrument
         """
-
+        warn("Do not call `Adapter.ask`, but `Instrument.ask` instead.",
+             FutureWarning)
         self.write(command)
-        if self.rw_delay is not None:
-            time.sleep(self.rw_delay)
+        self.wait_till_read()
         return self.read()
 
-    def write(self, command):
-        """ Writes the command to the GPIB address stored in the
-        :attr:`.address`
+    def write(self, command, **kwargs):
+        """Write a string command to the instrument appending `write_termination`.
+
+        If the GPIB address in :attr:`.address` is defined, it is sent first.
 
-        :param command: SCPI command string to be sent to the instrument
+        :param str command: Command string to be sent to the instrument
+            (without termination).
+        :param kwargs: Keyword arguments for the connection itself.
         """
-        if self.address is not None:
-            address_command = "++addr %d\n" % self.address
-            self.connection.write(address_command.encode())
-        command += "\n"
-        self.connection.write(command.encode())
+        if self.address is not None and not command.startswith("++"):
+            super().write("++addr %d" % self.address, **kwargs)
+        super().write(command, **kwargs)
 
     def _format_binary_values(self, values, datatype='f', is_big_endian=False, header_fmt="ieee"):
         """Format values in binary format, used internally in :meth:`.write_binary_values`.
 
+        .. deprecated:: 0.10
+            Implement the code in the instrument itself.
+
```

Comment:
I think this method could be part of the `Adapter` class since it implements standard formatting for binary data which is independent from particular instrument.
In addition, the escape code few lines below is specific to `Prologix` and it should not be moved at `Instrument` level.

Reply:
Oh sorry, I think I forgot to remove the depreciation warning. Exactly for the reasons you mentioned, I defined write binary (and format binary) at adapter level.